Add bookings feat: завершена реализация функционала бронирований (ФЗ-15)

cepheil opened the pull request:
Реализованы все требования ТЗ-15:
- Подготовлены SQL-схемы.
- Добавлен модуль бронирований:
  • Booking, BookingDto, BookingShortDto, BookerDto
  • BookingController, BookingServiceImpl, BookingRepository
  • BookingMapper для преобразования между моделями и DTO
- Поддержка подтверждения/отклонения бронирований владельцем вещи
- Добавлены даты бронирования при просмотре вещей
- Реализованы интеграции бронирований с пользователями и вещами
- Добавлены отзывы на вещь после того, как взяли её в аренду
- Добавлены мапперы и сервисы для работы с комментариями, запросами вещей, пользователями
- Обновлены контроллеры (User, Item, ItemRequest, Booking)
- Добавлены DTO для сокращённого отображения (ItemShortDto, ItemWithBookingsDto)

- Добавлен compose.yaml для локального запуска PostgreSQL

cepheil committed:
реализация ТЗ15 add-bookings

cepheil committed:
реализация ТЗ15 add-bookings

-BookingController
-BookingRepository
-BookingServiceImpl

cepheil committed:
feat: завершена реализация функционала бронирований (ФЗ-15)
Реализованы все требования ТЗ-15:
- Подготовлены SQL-схемы.
- Добавлен модуль бронирований:
  • Booking, BookingDto, BookingShortDto, BookerDto
  • BookingController, BookingServiceImpl, BookingRepository
  • BookingMapper для преобразования между моделями и DTO
- Поддержка подтверждения/отклонения бронирований владельцем вещи
- Добавлены даты бронирования при просмотре вещей
- Реализованы интеграции бронирований с пользователями и вещами
- Добавлены отзывы на вещь после того, как взяли её в аренду
- Добавлены мапперы и сервисы для работы с комментариями, запросами вещей, пользователями
- Обновлены контроллеры (User, Item, ItemRequest, Booking)
- Добавлены DTO для сокращённого отображения (ItemShortDto, ItemWithBookingsDto)

- Добавлен compose.yaml для локального запуска PostgreSQL

cepheil committed:
checkstyle fix

cepheil committed:
fix

cepheil committed:
fix

cepheil committed:
fix

cepheil committed:
fix

avfyodorov commented on src/main/java/ru/practicum/shareit/booking/Booking.java:
```diff
@@ -9,18 +11,38 @@
 import java.time.LocalDateTime;
 
 @Data
```
Data
Лучше не использовать эту аннотацию для сущностей. Дело в том, что ломбок генерирует EqualsAndHashCode, И могут быть конфликты с аналогичными механизмами hibernate’a
В данной работе ничего страшного не произойдет, но лучше не привыкать).
Если бы в проект добавили кэширование, то возникли бы проблемы.
https://stackoverflow.com/questions/75181366/why-jpa-buddy-complains-about-data-annotation-over-jpa-entity

avfyodorov commented on src/main/java/ru/practicum/shareit/item/model/Item.java:
```diff
@@ -1,26 +1,47 @@
+import ru.practicum.shareit.user.User;
 
 
 @Data
```
и здесь Data лучше заменить на Getter/Setter.
Посмотрите, пожалуйста,  аналогичные ситуации и в других сущностях.

avfyodorov commented on src/main/java/ru/practicum/shareit/booking/mapper/BookingMapper.java:
```diff
@@ -0,0 +1,61 @@
+import ru.practicum.shareit.user.User;
+
+@NoArgsConstructor(access = AccessLevel.PRIVATE)
+public class BookingMapper {
```
Принято максимально избавляться от шаблонного кода, чтобы не перегружать и так большие и объемные проекты. Для маппинга сущностей часто используется библиотека Mapstruct, которая позволяет автоматически настраивать маппинг из одной сущности в другую с помощью описания интерфейсов и аннотаций.
Вот здесь есть отличный гайд для старта работы с библиотекой.
https://web.archive.org/web/20240619130052/https://rukovodstvo.net/posts/id_799/

//  уже нет. только из архива  https://rukovodstvo.net/posts/id_799/

Что такое Mapstruct и как правильно настроить его для модульного тестирования в SpringBoot приложениях
https://javarush.com/groups/posts/3698-chto-takoe-mapstruct-i-kak-praviljhno-nastroitjh-ego-dlja-moduljhnogo-testirovanija-v-springboo

При наличии свободного времени можно поизучать и попробовать прикрутить в проект
Необязательно, просто для информации .

cepheil replied:
спасибо. обязательно изучу.

avfyodorov commented on src/main/resources/application-test.properties:
```diff
@@ -1,13 +1,50 @@
+# spring.jpa.hibernate.ddl-auto=none
+#spring.jpa.hibernate.ddl-auto=create-drop
+
+###########################################################################
```
Всё же лучше, по возможности, привести файл в порядок... убрать лишнее.

cepheil replied:
сделано.

avfyodorov reviewed (commented):
Добрый день, Сергей!

Практически всё сделано, осталось чуть доделать, поправить опечатки.

cepheil committed:
Refactor:
Заменены @Data на явные аннотации (@Getter, @Setter, @NoArgsConstructor, @AllArgsConstructor, @EqualsAndHashCode(onlyExplicitlyIncluded = true)).
Теперь equals и hashCode работают безопасно для JPA

avfyodorov reviewed (commented):
Добрый день, Сергей!

Замечаний нет.
Работа принята.
